### PR TITLE
[CROSSDATA-250]Globalindex

### PR DESCRIPTION
**Pull request by pianista215**
## Description
Global index functionality

https://stratio.atlassian.net/wiki/display/CROSSDATA/Introduction?src=contextnavpagetreemode
https://stratio.atlassian.net/wiki/display/CROSSDATA/Features
https://stratio.atlassian.net/wiki/display/CROSSDATA/Reference+Guide

### Testing
- [X] Unit, integration tests

### Documentation
- [X] Changelog update
- [X] Documentation link

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/crossdata/XDContext.scala [hunk 1]
```diff
@@ -70,6 +71,7 @@ class XDContext protected (@transient val sc: SparkContext,
   def this(sc: SparkContext, config: Config) =
     this(sc, Some(config))
 
+  import XDContext._
```
Comment (pfcoperez) [inline core/src/main/scala/org/apache/spark/sql/crossdata/XDContext.scala]: Either this import or ...

File: core/src/main/scala/org/apache/spark/sql/crossdata/XDContext.scala [hunk 2]
```diff
@@ -79,8 +81,7 @@ class XDContext protected (@transient val sc: SparkContext,
      Config should be changed by a map and implicitly converted into `Config` whenever one of its
      methods is called.
      */
-
-  import XDContext.{catalogConfig, config, xdConfig}
+  import XDContext.{catalogConfig, xdConfig}
```
Comment (pfcoperez) [inline core/src/main/scala/org/apache/spark/sql/crossdata/XDContext.scala]: ...this one should disappear. That is, the should coalesce.

File: core/src/main/scala/org/apache/spark/sql/crossdata/catalog/CatalogChain.scala [hunk 3]
```diff
@@ -125,16 +125,30 @@ private[crossdata] class CatalogChain private(val temporaryCatalogs: Seq[XDTempo
   override def persistView(tableIdentifier: ViewIdentifier, plan: LogicalPlan, sqlText: String): Unit =
     persistentCatalogs.foreach(_.saveView(tableIdentifier, plan, sqlText))
 
+  override def persistIndex(crossdataIndex: CrossdataIndex): Unit =
+    if (persistentChainedLookup(_.lookupTable(crossdataIndex.tableIdentifier)).isEmpty) {
```
Comment (pfcoperez) [inline core/src/main/scala/org/apache/spark/sql/crossdata/catalog/CatalogChain.scala]: You could call `tableMetadata` here so you can get updates of the table look-up at persistent catalog behavior

File: core/src/main/scala/org/apache/spark/sql/crossdata/catalog/CatalogChain.scala [hunk 4]
```diff
@@ -125,16 +125,30 @@ private[crossdata] class CatalogChain private(val temporaryCatalogs: Seq[XDTempo
...
+      persistentCatalogs.foreach(_.saveIndex(crossdataIndex))
+    }
+
+
   override def dropTable(tableIdentifier: TableIdentifier): Unit = {
     val strTable = tableIdentifier.unquotedString
     if (!tableExists(tableIdentifier)) throw new RuntimeException(s"Table $strTable can't be deleted because it doesn't exist")
     logInfo(s"Deleting table $strTable from catalog")
+
+    indexMetadataByTableIdentifier(tableIdentifier) map { index =>
```
Comment (pfcoperez) [inline core/src/main/scala/org/apache/spark/sql/crossdata/catalog/CatalogChain.scala]: Use `foreach` instead of `map` as you are interested just in the `dropIndex` side effect and not its return value.

File: core/src/main/scala/org/apache/spark/sql/crossdata/catalog/CatalogChain.scala [hunk 5]
```diff
@@ -152,6 +166,32 @@ private[crossdata] class CatalogChain private(val temporaryCatalogs: Seq[XDTempo
...
+  override def dropIndex(indexIdentifier: IndexIdentifier): Unit = {
+    val strIndex = indexIdentifier.unquotedString
+    if(indexMetadata(indexIdentifier).isEmpty) throw new RuntimeException(s"Index $strIndex can't be deleted because it doesn't exist")
+    logInfo(s"Deleting index ${indexIdentifier.unquotedString} from catalog")
+
+    //First remove table that holds the index
+    if(tableExists(indexIdentifier.asTableIdentifier))
+      dropTable(indexIdentifier.asTableIdentifier)
+
+    persistentCatalogs foreach(_.dropIndex(indexIdentifier))
```
Comment (pfcoperez) [inline core/src/main/scala/org/apache/spark/sql/crossdata/catalog/CatalogChain.scala]: Ok, you've verified that the index exists at least at one persistent level. What if, at some other level, it doesn't and its actual implementation raises an exception at `dropIndex`?

File: core/src/main/scala/org/apache/spark/sql/crossdata/catalog/CatalogChain.scala [hunk 6]
```diff
@@ -152,6 +166,32 @@ private[crossdata] class CatalogChain private(val temporaryCatalogs: Seq[XDTempo
...
+  }
+
+  override def indexMetadata(tableIdentifier: IndexIdentifier): Option[CrossdataIndex]=
+    persistentChainedLookup(_.lookupIndex(tableIdentifier))
+
+  override def indexMetadataByTableIdentifier(tableIdentifier: TableIdentifier):Option[CrossdataIndex]=
+    persistentCatalogs map (_.lookupIndexByTableIdentifier(tableIdentifier)) collectFirst {
```
Comment (pfcoperez) [inline core/src/main/scala/org/apache/spark/sql/crossdata/catalog/CatalogChain.scala]: There is a performance issue here: `map` will execute `lookupIndexByTableIdentifier` for every single persistent catalog. No matter at which position is located the first hit. You should use the `view` method: `persistentCatalogs.view map`

File: core/src/main/scala/org/apache/spark/sql/crossdata/catalog/persistent/DerbyCatalog.scala [hunk 7]
```diff
@@ -120,8 +145,8 @@ class DerbyCatalog(sqlContext: SQLContext, override val catalystConf: CatalystCo
       val optsJSON = resultSet.getString(OptionsField)
       val version = resultSet.getString(CrossdataVersionField)
 
-      Some(
-        CrossdataTable(table, Some(database), Option(deserializeUserSpecifiedSchema(schemaJSON)), datasource,
+      Option(
```
Comment (pfcoperez) [inline core/src/main/scala/org/apache/spark/sql/crossdata/catalog/persistent/DerbyCatalog.scala]: why the change?

File: core/src/main/scala/org/apache/spark/sql/crossdata/catalog/persistent/DerbyCatalog.scala [hunk 8]
```diff
@@ -155,6 +180,30 @@ class DerbyCatalog(sqlContext: SQLContext, override val catalystConf: CatalystCo
...
+      val database = resultSet.getString(DatabaseField)
+      val table = resultSet.getString(TableNameField)
+      val indexName = resultSet.getString(IndexNameField)
+      val indexType = resultSet.getString(IndexTypeField)
+      val indexedCols = resultSet.getString(IndexedColsField)
+      val pk = resultSet.getString(PKField)
+      val datasource = resultSet.getString(DatasourceField)
+      val optsJSON = resultSet.getString(OptionsField)
+      val version = resultSet.getString(CrossdataVersionField)
+
```
Comment (pfcoperez) [inline core/src/main/scala/org/apache/spark/sql/crossdata/catalog/persistent/DerbyCatalog.scala]: The `resultSet` is not being closed

File: core/src/main/scala/org/apache/spark/sql/crossdata/catalog/persistent/DerbyCatalog.scala [hunk 9]
```diff
@@ -202,7 +251,6 @@ class DerbyCatalog(sqlContext: SQLContext, override val catalystConf: CatalystCo
     }
 
   override def persistViewMetadata(tableIdentifier: TableIdentifier, sqlText: String): Unit =
-
     try {
       connection.setAutoCommit(false)
       val resultSet = selectMetadata(TableWithViewMetadata, tableIdentifier)
```
Comment (pfcoperez) [inline core/src/main/scala/org/apache/spark/sql/crossdata/catalog/persistent/DerbyCatalog.scala]: This isn't your fault but his `resultSet` is not being closed

Comment (pfcoperez) [inline core/src/main/scala/org/apache/spark/sql/crossdata/catalog/persistent/DerbyCatalog.scala]: ... Not `connection`

File: core/src/main/scala/org/apache/spark/sql/crossdata/catalog/persistent/DerbyCatalog.scala [hunk 10]
```diff
@@ -310,13 +416,59 @@ class DerbyCatalog(sqlContext: SQLContext, override val catalystConf: CatalystCo
     preparedStatement.setString(2, tableIdentifier.table)
     preparedStatement.executeQuery()
 
+  }
 
+  private def selectIndex(indexIdentifier: IndexIdentifier): ResultSet = {
+    val preparedStatement = connection.prepareStatement(s"SELECT * FROM $DB.$TableWithIndexMetadata WHERE $IndexNameField= ? AND $IndexTypeField= ?")
+    preparedStatement.setString(1, indexIdentifier.indexName)
+    preparedStatement.setString(2, indexIdentifier.indexType)
+    preparedStatement.executeQuery()
```
Comment (pfcoperez) [inline core/src/main/scala/org/apache/spark/sql/crossdata/catalog/persistent/DerbyCatalog.scala]: You should close `prepareStatement`

File: core/src/main/scala/org/apache/spark/sql/crossdata/catalog/persistent/DerbyCatalog.scala [hunk 11]
```diff
@@ -310,13 +416,59 @@ class DerbyCatalog(sqlContext: SQLContext, override val catalystConf: CatalystCo
...
+  private def tableSchemaExists(schema: String, table: String, connection: Connection): Boolean =  {
+    val query =
+      s"""|SELECT * FROM SYS.SYSSCHEMAS sch
+          |LEFT JOIN SYS.SYSTABLES tb ON tb.schemaid = sch.schemaid
+          |WHERE sch.SCHEMANAME='$schema' AND tb.TABLENAME='${table.toUpperCase}'""".stripMargin
+
+    val preparedStatement = connection.prepareStatement(query)
+    val resultSet = preparedStatement.executeQuery()
+
+    resultSet.next()
```
Comment (pfcoperez) [inline core/src/main/scala/org/apache/spark/sql/crossdata/catalog/persistent/DerbyCatalog.scala]: Do close `preparedStatement` & `resultSet`

File: core/src/main/scala/org/apache/spark/sql/crossdata/catalog/persistent/ZookeeperCatalog.scala [hunk 12]
```diff
@@ -186,7 +186,63 @@ class ZookeeperCatalog(sqlContext: SQLContext, override val catalystConf: Cataly
...
+  override def dropIndexMetadata(indexIdentifier: IndexIdentifier): Unit =
+    indexDAO.dao.getAll().filter(
+      index => index.crossdataIndex.indexIdentifier == indexIdentifier
+    ) foreach (selectedIndex => indexDAO.dao.delete(selectedIndex.indexId))
+
+  override def dropAllIndexesMetadata(): Unit =
+    indexDAO.dao.deleteAll
+
+  override def lookupIndex(indexIdentifier: IndexIdentifier): Option[CrossdataIndex] = {
+    if (indexDAO.dao.count > 0)
```
Comment (pfcoperez) [inline core/src/main/scala/org/apache/spark/sql/crossdata/catalog/persistent/ZookeeperCatalog.scala]: Something like:

```scala
  val res = indexDAO.dao.getAll().find(_.crossdataIndex.indexIdentifier == indexIdentifier) map(_.crossdataIndex)
    if(res.isEmpty) indexDAO.logger.warn("Index path doesn't exist")
    res
```

might be safer and easier to read

File: core/src/main/scala/org/apache/spark/sql/crossdata/catalog/persistent/ZookeeperCatalog.scala [hunk 13]
```diff
@@ -186,7 +186,63 @@ class ZookeeperCatalog(sqlContext: SQLContext, override val catalystConf: Cataly
...
+
+  }
+
+  override def dropIndexMetadata(tableIdentifier: TableIdentifier): Unit =
+    indexDAO.dao.getAll().filter(
+      index => index.crossdataIndex.tableIdentifier == tableIdentifier
+    ) foreach (selectedIndex => indexDAO.dao.delete(selectedIndex.indexId))
+
+  override def lookupIndexByTableIdentifier(tableIdentifier: TableIdentifier): Option[CrossdataIndex] = {
+    if (indexDAO.dao.count > 0)
```
Comment (pfcoperez) [inline core/src/main/scala/org/apache/spark/sql/crossdata/catalog/persistent/ZookeeperCatalog.scala]: Same as https://github.com/Stratio/Crossdata/pull/611/files#diff-5415d14c4cee08d017ab5538ce6d7435R210

File: core/src/main/scala/org/apache/spark/sql/crossdata/catalyst/globalindex/IndexUtils.scala [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    checkIfRemainExprAreSupported(Seq(condition))
+  }
+
+  /**
+    * Check if predicate is supported by ElasticSearch native queries to use the index
+    *
+    * @param predicate
+    * @return
+    */
+  def isSupportedPredicate(predicate: Predicate): Boolean = predicate match {
```
Comment (pfcoperez) [inline core/src/main/scala/org/apache/spark/sql/crossdata/catalyst/globalindex/IndexUtils.scala]: To prevent someone in the future from screwing it up by putting a `false` instead of a `true`:

```scala
def isSupportedPredicate(predicate: Predicate): Boolean = predicate match {
    //TODO: Add more filters?? Reference: ElasticSearchQueryProcessor
    case _: And |
         _ : Contains |
         _: EqualTo |
         _: GreaterThan |
         _: GreaterThanOrEqual |
         _: In |
         _: IsNull |
         _: IsNotNull |
         _: LessThan |
         _: LessThanOrEqual |
         _ : Or |
         _: StartsWith => true
    case _ => false
  }
```

File: core/src/main/scala/org/apache/spark/sql/crossdata/catalyst/optimizer/XDOptimizer.scala [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+  private def analyzeAndOptimize(plan: LogicalPlan): LogicalPlan = {
+    xdContext.optimizer.execute(analyze(plan))
+  }
+
+  private def schemaToAttribute(schema: StructType): Seq[UnresolvedAttribute] =
+    schema.fields map {field => UnresolvedAttribute(field.name)}
+
+  private def resultPksToLiterals(rows: Array[Row], dataType:DataType): Seq[Literal] = // TODOrow changed
+    rows map { row =>
+      val valTransformed = row.toSeq.head // TODO changed
```
Comment (pfcoperez) [inline core/src/main/scala/org/apache/spark/sql/crossdata/catalyst/optimizer/XDOptimizer.scala]: Too expensive, I'd rather use `val valTransformed = row.get(0)`

**Comment by pfcoperez** [comment]
+1

**Comment by darroyocazorla** [comment]
:+1:
